tools/perfdw.d

```diff
@@ -0,0 +1,95 @@
+/*
+ * Trace all IOs for each downstairs from the time they are sent over
+ * the network socket to when the ack for an IO is received.
+ * Group by IO type (R/W/F) and client ID (Which downstairs).
+ *
+ * arg0 is the job ID number.
+ * arg1 is the client ID
+ */
+cdt*:::gw_read_submit_start,
+cdt*:::gw_write_submit_start,
+cdt*:::gw_flush_submit_start
+/arg1 == 0/
+{
+    start0[arg0] = timestamp;
+}
+cdt*:::gw_read_submit_start,
+cdt*:::gw_write_submit_start,
+cdt*:::gw_flush_submit_start
+/arg1 == 1/
+{
+    start1[arg0] = timestamp;
+}
+cdt*:::gw_read_submit_start,
+cdt*:::gw_write_submit_start,
+cdt*:::gw_flush_submit_start
+/arg1 == 2/
+{
+    start2[arg0] = timestamp;
+}
+
+/*
+ * When a read ACK comes back, calculate the delta and store it
+ */
+cdt*:::gw_read_submit_end
```

Comment:
I think the aggregation for all timestamps could have two keys, one for the operation, one for the client ID. E.g., `@time["read", arg1] = ...`. For reference, [this section](https://illumos.org/books/dtrace/chp-aggs.html#chp-aggs-2) of the DTrace guide discusses aggregations with multiple keys, about 4 examples into the section.

You could also make the start times an associative array, indexed by the client ID and job ID: `start[arg1, arg0] = timestamp;`. Combined with the above, I think you can group all the start actions into one, and the three end actions for each operation type (r/w/f) into one per type.

@ahl Would definitely have some more concrete suggestions here too.

Reply:
Totally agreed, @bnaecker. `start[arg0, arg1] = timestamp` or whatever

Reply:
Ah, yes.  Much better

```
  cdt*:::gw_read_submit_start,
  cdt*:::gw_write_submit_start,
  cdt*:::gw_flush_submit_start
  {
      start[arg0, arg1] = timestamp;
  }

  cdt*:::gw_read_submit_end
  /start[arg0, arg1] != 0/
  {
      @time["read  client", arg1] = quantize(timestamp - start[arg0, arg1]);
      start[arg0, arg1] = 0;
  }

```

Still gives me a unique report for each client (though I do have to look to the left to see the client)
```
  read downstairs client                                            0
           value  ------------- Distribution ------------- count
          524288 |                                         0
         1048576 |@@@                                      3
         2097152 |                                         0
         4194304 |@@@@@                                    6
         8388608 |@@@@@@@@@@@                              12
        16777216 |@@@@@@@@@@@@@@@@@@@                      21
        33554432 |@@                                       2
        67108864 |                                         0
```